### PR TITLE
Design Doc Draft

### PR DESCRIPTION
**Pull request by ywwg**

### CODE REVIEW

File: proposals/2023-08-21-utf8.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+### prometheus/prometheus
+
+All support for UTF8 will be behind a default-off configuration flag. Behavior will not be changed unless this flag is on.
+
+Prometheus itself needs updates to its lexers and parsers to support reading the new exposition syntax and promql syntax:
+
+* openmetricslex
+* promlex
+* generated_parser
+
```
Comment (fionaliao) [inline proposals/2023-08-21-utf8.md:209]: Also https://github.com/prometheus/prometheus/blob/main/web/ui/module/codemirror-promql/src/parser/parser.ts (used for syntax highlighting, linting and autocompletion)

File: proposals/2023-08-21-utf8.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+We do not directly address the issue of adding function-like syntaxes, such as `metricname.sum`. The changes proposed here should be forward-compatible with such a new syntax, as long as the new syntax is outside of the quoted metric names.
+
+We do not propose a syntax for selecting metric names other than by basic equality. The changes proposeed here would be forward-compatible with ways of doing that, for instance operator prefixes on the quoted metric name.
+
+## How
+
+We propose the following change in syntax to PromQL and the exposition format:
+
+* Label names may be quoted
```
Comment (fionaliao) [inline proposals/2023-08-21-utf8.md]: Should this be more specific on what sort of quotes are supported? (double quotes, single quotes and backticks)

File: proposals/2023-08-21-utf8.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+The HTTP endpoints will need to support html unescaping metric names where they occur.
+
+## Alternatives
+
+### Non-quoted metric names
+
+`my.dotted.metric{label="value"}`
+
+While this syntax could be made to work with simple additional characters like dots, it does not fulfill the goal of full UTF8 support, which would include `{` or other special characters without a verbose and unreadable escaping mechanism. Selecting this approach would introduce a convenient syntax for only a small new subset of UTF8, so a quoting mechanism would still be needed for the full character set. Rather than implement this half-improvement we decided to only support a full quoting approach.
```
Comment (fionaliao) [inline proposals/2023-08-21-utf8.md]: Could you add an example of how the escaping would have to work for special characters?

File: proposals/2023-08-21-utf8.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+We do not directly address the issue of adding function-like syntaxes, such as `metricname.sum`. The changes proposed here should be forward-compatible with such a new syntax, as long as the new syntax is outside of the quoted metric names.
+
+We do not propose a syntax for selecting metric names other than by basic equality. The changes proposeed here would be forward-compatible with ways of doing that, for instance operator prefixes on the quoted metric name.
+
+## How
+
+We propose the following change in syntax to PromQL and the exposition format:
+
+* Label names may be quoted
+* A quoted string inside the label set (`{}`) without an operator is default-assigned to the label name `__name__` (the metric name).
```
Comment (fionaliao) [inline proposals/2023-08-21-utf8.md]: What would be the preferred way of writing metric and label names from the current allowed characters? With quotes or without quotes and with the metric name outside the braces? 

[In this comment thread](https://docs.google.com/document/d/1yFj5QSd1AgCYecZ9EJ8f2t4OgF2KBZgJYVde-uzVEtI/edit?disco=AAAAojYus7c) @beorn7 says:
> That's indeed the core idea here: As long as you go with the recommended character set, you get easy syntax. If you really have to use "weird" characters, you have to jump through syntactical hoops.
> That's not unheard of. Cf. quoting rules in Yaml, UNIX shell commands, …
> And remember that the intent here is to support a niche use case, but not to change the common use case.

So I assume the intent is to keep writing metrics in the current way when possible. I'm not sure how realistic the expectation is though of utf-8 staying a niche use case e.g. since `.` is commonly supported by other metric languages I suspect `.`s will start becoming used a lot more when it becomes possible to do so.

File: proposals/2023-08-21-utf8.md (hunk 3 above)
Comment (ywwg) [reply proposals/2023-08-21-utf8.md]: done.  I also fixed some of the syntax examples to match

File: proposals/2023-08-21-utf8.md (hunk 4 above)
Comment (ywwg) [reply proposals/2023-08-21-utf8.md]: it will be necessary to use the quoting syntax for anything more than traditional prometheus names.

To me, I see this as a move towards unifying promql with logql and alerting queries.  For logs and alerts, there is never anything on the outside of the brackets because they don't have the concept of a metric name. In this way, metrics are an exception with an odd outlier syntax. By bringing the name inside the brackets, metrics are now closer in syntax to the other languages.  The only difference is the convenience of the non-operator syntax which indicates that this is the special argument to the selector.

File: proposals/2023-08-21-utf8.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+We wish to use Prometheus as a backing database for other metrics implementations like Open Telemetry that use Prometheus-illegal characters, like '.'. Establishing a quoting syntax avoids the need for verbose and unreadable character encoding schemes, as well as non-roundtripable replacement schemes like using underscore for all unsupported characters.
+
+### Pitfalls of the current solution
+
+The current recommendation is to replace all unsupported characters with underscores. This substitution cannot be reversed, creating possible name collisions. Furthermore, the change in metric names between original write and eventual read can be confusing for end users.
+
+## Goals
+
+* Allow arbitrarily-named metric and label names to be submitted to Prometheus.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: Allow arbitrary metric and label names...

or

Allow arbitrarily-named metrics and labels...

File: proposals/2023-08-21-utf8.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+* **Implementation Status:** `Proof-of-concept implemented`
+
+* **Related Issues and PRs:**
+  * [GH Issue](https://github.com/prometheus/prometheus/issues/12630)
+  * [PR](https://github.com/grafana/mimir-prometheus/pull/476) (TODO needs to be moved)
+
+* **Other docs or links:**
+  * [Background Discussion / Justification](https://docs.google.com/document/d/1yFj5QSd1AgCYecZ9EJ8f2t4OgF2KBZgJYVde-uzVEtI/edit).  Please read this document first for more information on the chosen solution.
+
+> TL;DR: We propose to add support for the UTF8 characters in Prometheus metric and label names. While tsdb and the Prometheus code already support arbitrary strings, PromQL and the exposition format need a quoting syntax to make these values distinguishable.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: Nit: UTF8 → UTF-8 (throughout the doc)

File: proposals/2023-08-21-utf8.md (hunk 4 above)
Comment (beorn7) [reply proposals/2023-08-21-utf8.md]: It should be noted that Prometheus is the de-facto standard in cloud-native monitoring. It is reasonable to assume that other monitoring systems using non-Prometheus compliant names will become more and more of a niche rather than more common. It is definitely a non-goal to encourage native Prometheus users to use `.` more, as we have decade-old plans to use `.` as an operator in PromQL.

The notable exception to all of that is OTel, as it is the new and upcoming thing rather than the old and declining thing. However, the problem with OTel isn't just the `.`, it allows _all_ UTF-8 characters, and some of them even quite commonly. If people deal with OTel names, they generally have to use some quoting or escaping schema anyway, even if PromQL abandoned its plans to use `.` as an operator. Or in other words: By simply adding `.` to the list of allowed characters, we would only solve half of the problem but still pay the cost of losing `.` as an operator.

File: proposals/2023-08-21-utf8.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+We do not directly address the issue of adding function-like syntaxes, such as `metricname.sum`. The changes proposed here should be forward-compatible with such a new syntax, as long as the new syntax is outside of the quoted metric names.
+
+We do not propose a syntax for selecting metric names other than by basic equality. The changes proposeed here would be forward-compatible with ways of doing that, for instance operator prefixes on the quoted metric name.
+
+## How
+
+We propose the following change in syntax to PromQL and the exposition format:
+
+* Label names may be double-quoted.
+* A quoted string inside the label set (`{}`) without an operator is default-assigned to the label name `__name__` (the metric name).
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: Let's better use "label matching operator" instead of just "operator" (here and below) because the general term "operator" is more connected to what's described on https://prometheus.io/docs/prometheus/latest/querying/operators/ .

File: proposals/2023-08-21-utf8.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+* Style prefers, but does not enforce, that the operatorless string is the first term in the label set.
+* Strings in quotes will be unquoted using Go strconv.Unquote, so Go escaping can be used.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: But not if it backticks (which is the same as we have now for label values).

File: proposals/2023-08-21-utf8.md [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+* In the exposition format, the TYPE, HELP, and UNIT lines can have quotes around the metric name and similarly use Go quoting.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: IIRC the exposition format does not use Go quoting for label values right now. We should keep things consistent within the exposition format, i.e. quoted label and metric names should work exactly as label values.

File: proposals/2023-08-21-utf8.md [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+* Quoted label name: `{"my.dotted.metric", "http.status"=~"5.."}`
+
+* Classic metrics do not need to be treated differently:
+`sum(rate(regular_metric{region="east"}[5m]))`
+
+* If the user decides to put a classic metric inside the selector, it still needs to be quoted:
+`sum(rate({"regular_metric", region="east"}[5m]))`
+
+* Escape syntax if the metric has a quote: `sum(rate({"my \"quoted\" metric", region="east"}[5m]))`.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: Or use single quotes: `sum(rate({'my "quoted" metric', region="east"}[5m]))`

File: proposals/2023-08-21-utf8.md [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+
+* Recording Rule Syntax: `sum(rate({"my.dotted.metric:errors:rate5m", region="east"}[5m]))`. Recording rule syntax is a naming convention, not a programmatically-meaningful operator syntax.
+
+* Unicode escaping: `{"emoji metric: 💩 or \u1F4A9"}`
+
+* Possible future Native Histogram syntaxes (not being proposed here):
+  `{"my.metric.with.periods", region="east"}.sum`
+  `{"my.metric.with.periods".sum, region="east"}`
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: Not sure if the latter should be possible. (Not that important as it isn't proposed here, but maybe don't mention the 2nd line right now to avoid irrelevant discussions.)

To digress further: I think the usual case should be `my_histogram{foo="bar"}.sum`, not `my_histogram.sum{foo="bar"}`, i.e. add the `.` operator _after_ the selector, which would lead to the 1st line rather than the 2nd.

File: proposals/2023-08-21-utf8.md [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+
+* Possible future metric name operator syntax via prefixes (not being proposed here):
+  `{~"a\.regex.*selector.*", region="east"}`
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: Maybe add `{!"not.this.name", region="east"}` ?

File: proposals/2023-08-21-utf8.md [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+### Implementation
+
+Implementing this feature will require updates to several Prometheus libraries. These updates will be written so that functionality does not change for existing users and builds, and will only be enabled when certain flags are flipped or options turned on.  There will be minimal read-only backwards compatibility for clients that do not support the new supported character set and syntax.
+
+Main tasks:
+
+* Add flags / options to libraries and binaries to enable the new behavior.
+* Change / update code locations that check for a valid metric name, based on the flag setting.
+* Update text generators to generate the new syntax when a metric must be quoted.
+* Update parsers to parse the new syntax.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: This should distinguish between exposition format and PromQL (which only look similar, but are very different things).

File: proposals/2023-08-21-utf8.md [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+* Add an HTTP header so that clients can indicate their support for UTF8.
+* Add a way to escape metric names for clients that do not support the full character set.
+* Update endpoint code to support html-unquoting metric names in http API calls.
+
+### prometheus/common
+
+We will add a boolean argument to model.IsValidMetricName that switches validation logic to allow UTF8 names (and therefore will only check for 0 length). Because this is a public function, all users of this library will need to add an argument to their call sites.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: There are byte streams that are not valid UTF-8. That still has to be tested. Cf. how we validate label values right now.

File: proposals/2023-08-21-utf8.md [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+
+Other repositories that use this check to filter out "bad" metric names will have to be optional via a flag or argument.
+
+### prometheus/client_golang
+
+The golang library checks that metrics have a valid name when NewDesc is called. We can create a new parallel function, NewDescUTF8 (or similar) that will bypass the validity check.
+
+### prometheus/prometheus
+
+All support for UTF8 will be behind a default-off configuration flag. Behavior will not be changed unless this flag is on.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: configuration flag → feature flag

File: proposals/2023-08-21-utf8.md [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+Prometheus itself needs updates to its lexers and parsers to support reading the new exposition syntax and promql syntax:
+
+* openmetricslex (openmetrics exposition format)
+* promlex (prometheus exposition format)
+* generated_parser (PromQL parsing)
+* parser.ts (validation for text input fields)
+
+#### Backwards-compatibility
+
+Prometheus will also need to support queries and responses from old clients that do not support the new syntax. We will create a new boolean HTTP header called AllowUTF8Names. When this value is true in the request, and the feature is enabled in prometheus, UTF8 support will be enabled. For any client that does not pass the AllowUTF8Names header, any HTTP request that can return metric names (including simple queries) will have the UTF8 metric names escaped.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: This needs different treatment for exposition and the HTTP query API. The latter could use the HTTP header as described, while the former already has content negotiation and will just negotiate a newer version of OpenMetrics (or the classic text format, if we want to add the support to it). Protobuf will have no problem with encoding/decoding but we should still allow content negotiation via a version so that a scraper without UTF-8 support won't get presented with special characters in names.

File: proposals/2023-08-21-utf8.md [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+
+#### Text escaping
+
+We propose an escaping syntax for metric names that contain UTF8 characters:
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: This is for the use case of supporting layers in the stack that do not expect UTF-8 characters in names. (That isn't clear from just reading this section.)

File: proposals/2023-08-21-utf8.md [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+
+* Prefix the string with `U__`
+* All UTF8 characters will be encoded as underscores surrounding the unicode value, like `_1F60A_`
+* All pre-existing underscores will become doubled: `__`
+* If a string should start with "U__" already, it will need to be escaped: `U___55_____`. (That's `U__` + `_55_` + `__` + `__`)
+
+#### HTTP escaping
+
+The HTTP endpoints will need to support html unescaping metric names where they occur.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: Could you explain this point more? Why is HTML suddenly in the game?

File: proposals/2023-08-21-utf8.md [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+
+Main tasks:
+
+* Add flags / options to libraries and binaries to enable the new behavior.
+* Change / update code locations that check for a valid metric name, based on the flag setting.
+* Update text generators to generate the new syntax when a metric must be quoted.
+* Update parsers to parse the new syntax.
+* Add an HTTP header so that clients can indicate their support for UTF8.
+* Add a way to escape metric names for clients that do not support the full character set.
+* Update endpoint code to support html-unquoting metric names in http API calls.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: Why HTML?

File: proposals/2023-08-21-utf8.md (hunk 9 above)
Comment (beorn7) [reply proposals/2023-08-21-utf8.md]: While the text-based exposition formats (classic and OM) look similar to PromQL, they are very different things. We need more explanation here.

The already valid PromQL trick `{__name__="my.metric"}` doesn't work in the text-based exposition format. We have to state clearly, that we still want to allow `{"my.metric"}` in the exposition format (to keep the syntax similarities).

Quoting rules are different in the exposition format (see previous comment). Among others, only double quotes are allowed, no single quotes or backticks.

Finally, protobuf doesn't have any of these problems, but we still need some kind of content negotiation, see below.

File: proposals/2023-08-21-utf8.md [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+
+While this syntax could be made to work with simple additional characters like dots, it does not fulfill the goal of full UTF8 support, which would include `{` or other special characters without a verbose and less-readable escaping mechanism, e.g.
+
+  `my.dotted.metric\{with\ brackets\ and\ spaces\}{foo="bar"}`.
+
+Versus our selected approach:
+
+  `{"my.dotted.metric{with brackets and spaces}", foo="bar"}`
+
+Selecting this approach would introduce a convenient syntax for only a small new subset of UTF8, so a quoting mechanism would still be needed for the full character set. Rather than implement this half-improvement we decided to only support a full quoting approach.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: Also, the dot has been reserved as a potential operator for PromQL (with the histogram use case being on the table right now).

File: proposals/2023-08-21-utf8.md [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+
+Single, double, and backtick quotes are nearly equivalent in PromQL so this would just be the same as the above approach, or require changing the meaning of a particular quote character in a way that would break backward-compatibility. (Backticks prevent processing of escape sequences, which is otherwise enabled, see [PromQL documentation](https://prometheus.io/docs/prometheus/latest/querying/basics/#string-literals) for details.)
+
+For new users, it might be confusing to parse this syntax visually, with a bracket in the middle of a series of quoted values. The previous syntax resembles a function call, the way that `foo{bar}` is like `foo(bar)`. By adding the additional quotes this frame of reference is lost. It seems more intuitive to think of the brackets as **the characters that surround the parameters for selecting data**, where the "metric name" is one special case of that kind of selector. This is similar to the approach chosen for [LogQL](https://grafana.com/docs/loki/latest/query/) and [TraceQL](https://grafana.com/docs/tempo/latest/traceql/), which do not have a special case "metric name" equivalent and thus have no items outside the brackets.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: Looking at https://en.wikipedia.org/wiki/Bracket (and my own experience), it is confusing calling `{}` just "brackets" as that's a generic term in British English and commonly only referring to `[]` in American English (and `[]` also have a role in PromQL). Let's be specific and call them "curly brackets" or even less ambiguous "curly braces".

Comment (beorn7) [reply proposals/2023-08-21-utf8.md]: This paragraph sounds more like a justification of the chosen syntax rather than an explanation why backticks have been dismissed.

Maybe move this line of explanation more towards the top of the whole document?

File: proposals/2023-08-21-utf8.md [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+### VictoriaMetrics approach
+
+VictoriaMetrics has extended PromQL for UTF8 Support, calling the updated language [MetricsQL](https://docs.victoriametrics.com/MetricsQL.html):
+
+> * Metric names and label names may contain any unicode letter. For example температура{город="Киев"} is a value MetricsQL expression.
+> * Metric names and labels names may contain escaped chars. For example, foo\-bar{baz\=aa="b"} is valid expression. It returns time series with name foo-bar containing label baz=aa with value b. Additionally, the following escape sequences are supported:
+>   * \xXX, where XX is hexadecimal representation of the escaped ascii char.
+>   * \uXXXX, where XXXX is a hexadecimal representation of the escaped unicode char.
+
+By allowing non-escaped dots, they risk colliding with planned Native Histogram function calls (like "`.count`"). Or they might require those to be escaped, resulting in many backslashes as in our own non-quoted rejected approach.  We do plan to adopt the \u approach via unquoting as described above.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: IIUC they allow any character as non-escaped that currently has no special meaning in PromQL. So _any_ new feature in PromQL that requires a special character having a special meaning would be a breaking change for them. I don't think we have to stress the `.` operator thing again. (People like to come back to it, repeatedly ignoring that it only solves a part of the problem. By bringing up this special case over and over again, we enforce the notion that the only problem is the `.`.)

We should also mention that the ViMe approach leads to a lot of potential escaping, especially given the rise of OTel. An OTel user might very unsuspectingly use a name like `foo/bar/my-metric`, which becomes `foo\/bar\/my\-metric` in ViMe.

Finally, I wouldn't call it "we plan to adopt the \u approach" as the \u approach already works in label values, and we just keep the quoting rules consistent when allowing quoting of names. So it should be more like: "Note that the \u approach also works for the approach proposed here."

**Comment by beorn7** [comment]
Just realized that the little trick how to make templating work should also be in here.

**Review by beorn7 — CHANGES_REQUESTED** [review]

File: proposals/2023-08-21-utf8.md (hunk 22 above)
Comment (ywwg) [reply proposals/2023-08-21-utf8.md]: ah, missed this note, I'll apply this change

File: proposals/2023-08-21-utf8.md [hunk 23]
```diff
@@ -0,0 +1,245 @@
...
+* **Implementation Status:** `Proof-of-concept implemented`
+
+* **Related Issues and PRs:**
+  * [GH Issue](https://github.com/prometheus/prometheus/issues/12630)
+  * [PR](https://github.com/grafana/mimir-prometheus/pull/476) (TODO needs to be moved)
+
+* **Other docs or links:**
+  * [Background Discussion / Justification](https://docs.google.com/document/d/1yFj5QSd1AgCYecZ9EJ8f2t4OgF2KBZgJYVde-uzVEtI/edit).  Please read this document first for more information on the chosen solution.
+
+> TL;DR: We propose to add support for the UTF-8 characters in Prometheus metric and label names. While tsdb and the Prometheus code already support arbitrary strings, PromQL and the exposition format need a quoting syntax to make these values distinguishable.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: Maybe change "add support for the UTF-8 characters" into "add support for arbitrary UTF-8 characters"?

Also "support arbitrary strings" → "support arbitrary UTF-8 strings"

File: proposals/2023-08-21-utf8.md (hunk 22 above)
Comment (ywwg) [reply proposals/2023-08-21-utf8.md]: done

File: proposals/2023-08-21-utf8.md [hunk 24]
```diff
@@ -0,0 +1,245 @@
...
+`sum(rate(regular_metric{region="east"}[5m]))`
+
+* If the user decides to put a classic metric inside the selector, it still needs to be quoted:
+`sum(rate({"regular_metric", region="east"}[5m]))`
+
+* Escape syntax if the metric has a quote: `sum(rate({"my \"quoted\" metric", region="east"}[5m]))`.
+
+* Recording Rule Syntax: `sum(rate({"my.dotted.metric:errors:rate5m", region="east"}[5m]))`. Recording rule syntax is a naming convention, not a programmatically-meaningful operator syntax.
+
+* Unicode escaping: `{"\u1F4A9"}`
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: Is that a real unicode escaping? (2½ bytes?). Also used elsewhere.

File: proposals/2023-08-21-utf8.md [hunk 25]
```diff
@@ -0,0 +1,245 @@
...
+### Syntax Changes
+
+We propose the following change in syntax to PromQL and the exposition format:
+
+* Label names may be quoted, either double quotes for the exposition format, or singles/doubles/backticks for PromQL.
+* A quoted string inside the label set (`{}`) without a label matching operator is default-assigned to the label name `__name__` (the metric name).
+* Two strings without a label matching operator is a syntax error.
+* Style prefers, but does not enforce, that the string without a label matching operator is the first term in the label set.
+* Strings in quotes will be unquoted using the mechanism for each language: strconv.Unquote for PromQL, and the OpenMetrics escaping mechanism for the exposition format.
+* In the exposition format, the TYPE, HELP, and UNIT lines can have double quotes around the metric name and similarly use Go quoting.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: I think we should keep the quoting here consistent with the other quoting rules in OM (rather than suddenly introducing Go quoting). So this should also be "the OpenMetrics escaping mechanism".

File: proposals/2023-08-21-utf8.md [hunk 26]
```diff
@@ -0,0 +1,245 @@
...
+
+* Compatible with possible future Native Histogram syntax (not being proposed here):
+  `{"my.metric.with.periods", region="east"}.sum`
+
+* Compatible with possible future metric name operator syntax via prefixes (not being proposed here):
+  `{~"a\.regex.*selector.*", region="east"}` or `{!"not.this.name", region="east"}`
+
+### Unusual / Complex Characters
+
+Depending on the parsing implementation, there may be UTF-8 characters that, when encountered as native values in quotes, may cause problems with the lexers and parsers. These characters could include whitespace, especially newlines, or complex combiner characters. If such characters exist, we will require that they be escaped with `\u` rather than allow them to exist literally. As development of this proposal progresses, we will discover which characters will fall into this category.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: There is probably a difference between the exposition format (which is line based, so newlines have a special meaning, and (most likely) must not show up literally anywhere in names) and PromQL (where I could imagine newlines to just work). Not sure if it worth mentioning this here.

File: proposals/2023-08-21-utf8.md [hunk 27]
```diff
@@ -0,0 +1,245 @@
...
+### Templating syntax
+
+Templating is used in alerts, and for dashboards that want to reference metric values. Go templating already supports quoting by way of the `index` keyword. Whereas labels are usually denoted as `{{ $labels.my_label_name }}`, the quote syntax is `{{ index $labels "my.label.name" }}`. No code changes are required.
+
+## Backwards-compatibility
+
+There a few situations where we need to consider backwards compatibility:
+
+1. Old metrics provider, new backend: Old providers will not serve UTF-8 metric names, so nothing needs to be done here.
+2. New metrics provider, old backend: Old versions of Prometheus will throw an error if they see the new syntax. Therefore we must provide a way for the agent to serve previously-invalid metrics in a form that the old database can process.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: agent → scraped target

File: proposals/2023-08-21-utf8.md [hunk 28]
```diff
@@ -0,0 +1,245 @@
...
+4. Old backend, new query client: as with the first case, nothing needs to be done.
+
+### Text escaping
+
+We propose an escaping syntax for metric names that contain UTF-8 characters for systems that do not support UTF-8. This system can be used in queries and responses to provide compatibility between the systems.
+
+* Prefix the string with `U__`.
+* All non-valid characters (i.e. characters other than letters, numbers, and underscores) will be encoded as underscores surrounding the unicode value, like `_1F60A_`.
+* All pre-existing underscores will become doubled: `__`.
+* If a string should start with "U__" already, it will need to be escaped: `U___55_____`. (That's `U__` + `_55_` (for 'U') + `__` + `__`).
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: I have realized now that this approach doesn't really solve all the edge cases.

If there is an existing metric name that starts with `U__`, for example `U___263a_`, and we upgrade the backend from old to new, an old querier will suddenly receive the escaped version of the existing metric name, i.e. `U___55_______263a__` rather than the old form. Or in other words: We suddenly get a different response after an upgrade that is supposed to be compatible.

We could rule that we only apply the escaping where needed, i.e. only to newly appearing names with formerly invalid characters. But that would lead to another problem: Now a new metric called `☺` would be returned to an old client as `U___263a_`, while the existing metric would be returned in the same way.

The question is how relevant this is in practice. Is there an existing name that starts with `U___` (i.e. triple underscore) and then happens to have a valid unicode hex number after it and then another underscore?

However, the problem needs to be acknowledged here.

**Review by beorn7 — CHANGES_REQUESTED** [review]
Just a few nits, but also a couple of substantial comments. Most tricky is the one about the escaping schema.

File: proposals/2023-08-21-utf8.md (hunk 28 above)
Comment (ywwg) [reply proposals/2023-08-21-utf8.md]: > We could rule that we only apply the escaping where needed, i.e. only to newly appearing names with formerly invalid characters. 

Yes that is the approach I would propose

> But that would lead to another problem: Now a new metric called `☺` would be returned to an old client as `U___263a_`, while the existing metric would be returned in the same way.

I would content that  `U___263a_` and `☺` are the same metric, just two different ways of representing it. I see your point about upgrading the backend -- a query for `☺` will not hit `U___263a_`, because the backend will look for the native character. Perhaps that means the on-disk representation needs a flag in the metadata file to indicate if the block is UTF-8-native.  The library reading the tsdb will understand that a query for `☺` against blocks that do not have support should look for  `U___263a_`. The same query, hitting a newer block, will have  `☺`.

This may not be common but this is an excellent point, we need to cover this situation or else Bad Things will happen.  (And asking people to recode all of their persisted storage is not feasible)

File: proposals/2023-08-21-utf8.md (hunk 24 above)
Comment (ywwg) [reply proposals/2023-08-21-utf8.md]: I got that from here: https://www.compart.com/en/unicode/U+1F4A9 :).  I think it's implied to be \u01F4A9 ?

File: proposals/2023-08-21-utf8.md (hunk 28 above)
Comment (beorn7) [reply proposals/2023-08-21-utf8.md]: Maybe it's a good point to just claim that a pre-existing metric with the silly name `U___263a_` is just retconned as `☺` – as if it was always meant to be a smiley. :)

Comment (beorn7) [reply proposals/2023-08-21-utf8.md]: As said, all of this is fairly irrelevant in practice, but we have to formulate clear rules (maybe not complete rules in the design doc, but we need to point to the areas where the rules need to be completed eventually).

Comment (beorn7) [reply proposals/2023-08-21-utf8.md]: A remotely similar case was https://github.com/prometheus/prometheus/issues/5861 . The [test case](https://github.com/prometheus/prometheus/blob/83891135c6585fc2e9990824c9d12478acfe6b10/scrape/scrape_test.go#L1563-L1574) illustrates funny edge cases…

File: proposals/2023-08-21-utf8.md (hunk 24 above)
Comment (beorn7) [reply proposals/2023-08-21-utf8.md]: I don't think so. `\u` has to be followed by exactly four hex digits. `\U` has to be followed by exactly 8. Maybe for pile of poo, the only way in Go escaping is https://go.dev/play/p/W9R1pxipjEA

Comment (beorn7) [reply proposals/2023-08-21-utf8.md]: I guess what you used is the HTML entity, and maybe that really allows to skip leading zeros. My Go example uses `\U0001F4A9`, which boils down to `1F4A9` if leaving out the leading zeros.

Maybe let's use `\u263a` as the example character. It's shorter and doesn't need the UTF-32 edge case.

File: proposals/2023-08-21-utf8.md (hunk 28 above)
Comment (ywwg) [reply proposals/2023-08-21-utf8.md]: We could bump the version number in meta.json: https://ganeshvernekar.com/blog/prometheus-tsdb-persistent-block-and-its-index/#1-metajson

older versions would indicate that U__ metrics should be unescaped.  newer versions would expect native UTF-8 characters

Comment (ywwg) [reply proposals/2023-08-21-utf8.md]: code ref: https://github.com/prometheus/prometheus/blob/main/tsdb/block.go#L240

File: proposals/2023-08-21-utf8.md [hunk 29]
```diff
@@ -0,0 +1,258 @@
...
+
+Instead, we believe that it is more intuitive and easier to teach new users to think of the brace as **the characters that surround the parameters for selecting data**, where **the "metric name" is a special case** of that kind of selector. This is similar to the approach chosen for [LogQL](https://grafana.com/docs/loki/latest/query/) and [TraceQL](https://grafana.com/docs/tempo/latest/traceql/), which do not have a special case "metric name" equivalent and thus have no items outside the curly braces.
+
+Furthermore, PromQL users will already know that it's possible to put the metric name inside the braces, for the purpose of using an operator to enable richer metric selection, like: `{__name__=~"my\.metric.*"}`. The chosen syntax serves as a shortening of this operation, and is forward-compatible with future metric matching prefix operators if we decide to implement them.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: `{__name__=~"my\.metric.*"}` results in `parse error: unknown escape sequence U+002E`. Since PromQL query are first run through Go escaping (as explained in this doc ;), this has to be `{__name__=~"my\\.metric.*"}` or alternatively ``{__name__=~`my\.metric.*`}``.

File: proposals/2023-08-21-utf8.md [hunk 30]
```diff
@@ -0,0 +1,258 @@
...
+* All pre-existing underscores will become doubled: `__`.
+* If a string should start with "U__" already, it will need to be escaped: `U___55_____`. (That's `U__` + `_55_` (for 'U') + `__` + `__`).
+
+#### Mixed-Block Scenario 
+
+We must consider an edge case, in which a newer client persists metrics to disk in an older database that does not support UTF-8. Those metrics will be written to disk with the U__ escaping format. If, later, the user upgrades their database software, new metrics will be written with the native UTF-8 characters. At query time, there will be a problem: newer blocks were written with UTF-8 and older blocks were written with the escaping format. The query code will not know which encoding to look for.
+
+To avoid this confusion we propose to bump the version number in the tsdb meta.json file. On a per-block basis we will then know whether to look for UTF-8 metric names in native encoding (bumped version number 2) or in U__-escaped naming (current version 1).
+
+If a metric is written as "U__" but does not pass the unescaping algorithm, then we assume it was meant to be read as-is and let it pass through.
```
Comment (beorn7) [inline proposals/2023-08-21-utf8.md]: I think this is fine now, but it should still call out the remaining edge case. Maybe add something like:

"In turn, this means that an old metric or label name containing a valid escape sequence by chance would be parsed differently after enabling UTF-8 support. For example, a metric called `U___263a_` would appear as `☺` after the upgrade (if read from an old storage chunk, scraped from an old target, or received from an old remote-write client). This is an extremely unlikely scenario in practice so that the theoretically possible implicit name change is acceptable."

**Review by beorn7 — COMMENTED** [review]
Two more comments. Looks good otherwise.

Feel free to turn this into a PR against prometheus/proposals once you have addressed my comments as you see fit.

File: proposals/2023-08-21-utf8.md (hunk 30 above)
Comment (ywwg) [reply proposals/2023-08-21-utf8.md]: adopted as-is
